src/vendor/CowWrapper.sol

```diff
@@ -0,0 +1,163 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity >=0.7.6 <0.9.0;
+pragma abicoder v2;
+
+/// @title CoW Protocol v2 Authentication Interface
+/// @author CoW DAO developers
+interface CowAuthentication {
+    /// @dev determines whether the provided address is an authenticated solver.
+    /// @param prospectiveSolver the address of prospective solver.
+    /// @return true when prospectiveSolver is an authenticated solver, otherwise false.
+    function isSolver(address prospectiveSolver) external view returns (bool);
+}
+
+/// @title CoW Protocol Settlement Interface
+/// @notice Minimal interface for CoW Protocol's settlement contract
+/// @dev Used for type-safe calls to the settlement contract's settle function
+interface CowSettlement {
+    /// @notice Trade data structure matching GPv2Settlement
+    struct CowTradeData {
+        uint256 sellTokenIndex;
+        uint256 buyTokenIndex;
+        address receiver;
+        uint256 sellAmount;
+        uint256 buyAmount;
+        uint32 validTo;
+        bytes32 appData;
+        uint256 feeAmount;
+        uint256 flags;
+        uint256 executedAmount;
+        bytes signature;
+    }
+
+    /// @notice Interaction data structure for pre/intra/post-settlement hooks
+    struct CowInteractionData {
+        address target;
+        uint256 value;
+        bytes callData;
+    }
+
+    /// @notice Returns the authentication contract used by the settlement contract.
+    function authenticator() external returns (CowAuthentication);
+
+    /// @notice Settles a batch of trades atomically
+    /// @param tokens Array of token addresses involved in the settlement
+    /// @param clearingPrices Array of clearing prices for each token
+    /// @param trades Array of trades to execute
+    /// @param interactions Array of three interaction arrays (pre, intra, post-settlement)
+    function settle(
+        address[] calldata tokens,
+        uint256[] calldata clearingPrices,
+        CowTradeData[] calldata trades,
+        CowInteractionData[][3] calldata interactions
+    ) external;
+}
+
+/// @title CoW Protocol Wrapper Interface
+/// @notice Interface for wrapper contracts that add custom logic around CoW settlements
+/// @dev Wrappers can be chained together to compose multiple settlement operations
+interface ICowWrapper {
+    /// @notice A human readable label for this wrapper. Used for display in explorer/analysis UIs
+    function name() external view returns (string memory);
+
+    /// @notice Initiates a wrapped settlement call
+    /// @dev This is the entry point for wrapped settlements. The wrapper will execute custom logic
+    ///      before calling the next wrapper or settlement contract in the chain.
+    /// @param settleData ABI-encoded call to CowSettlement.settle()
+    /// @param wrapperData Encoded chain of wrapper-specific data followed by addresses of next wrappers/settlement
```

Comment:
Do you think "chain of wrapper-specific data followed by addresses of next wrappers/settlement" is self-explanatory? 

```suggestion
    /// @param wrapperData Sequence of pairs of wrapper-specific data and the addresses of the next wrapper that will interpret the data. The last wrapper will be responsible for calling the settlement contract. The bytes structure looks like `<wrapper-data-1><wrapper-address-1><wrapper-data-2><wrapper-address-2>...<wrapper-data-N><wrapper-address-N>`
```

Also, you add "settlement" here, but I think you removed that behaviour. Last item is not the address of the settlement, no?